reader.go

```diff
@@ -55,67 +100,200 @@ const (
+	}
+
+	genFetcher, err := newGenerationFetcher(
+		config.Session,
+		config.ClusterStateTracker,
+		readFrom,
+		config.Logger,
+	)
+	if err != nil {
+		return nil, err
 	}
 
 	reader := &Reader{
-		config:  &*config, // make a copy
-		stopped: 0,
+		config:     config.Copy(),
+		genFetcher: genFetcher,
+		readFrom:   readFrom,
+		stoppedCh:  make(chan struct{}),
 	}
 	return reader, nil
 }
 
 // Run runs the CDC reader. This call is blocking and returns after an error occurs, or the reader
 // is stopped gracefully.
-func (r *Reader) Run() error {
-	gen, err := r.findLatestGeneration()
-	if err != nil {
-		return err
-	}
+func (r *Reader) Run(ctx context.Context) error {
+	l := r.config.Logger
 
-	split, err := r.splitStreams(gen.streams)
-	if err != nil {
-		return err
-	}
+	runErrG, runCtx := errgroup.WithContext(ctx)
 
-	errG, ctx := errgroup.WithContext(r.config.Context)
-	for _, _group := range split {
-		group := _group
-		errG.Go(func() error {
-			return r.processStreamGroup(ctx, group)
-		})
-	}
+	runErrG.Go(func() error {
+		select {
+		case <-runCtx.Done():
+			return runCtx.Err()
+		case <-r.stoppedCh:
+		}
+		r.genFetcher.Stop()
+		return nil
+	})
+	runErrG.Go(func() error {
+		return r.genFetcher.Run(runCtx)
+	})
+	runErrG.Go(func() error {
+		gen, err := r.genFetcher.Get(runCtx)
+		if gen == nil {
+			return err
+		}
+
+		if r.readFrom.Before(gen.startTime) {
+			r.readFrom = gen.startTime
+		}
+
+		for {
+			l.Printf("starting reading generation %v from timestamp %v", gen.startTime, r.readFrom)
 
-	return errG.Wait()
+			if err := r.config.ProgressManager.StartGeneration(gen.startTime); err != nil {
+				return err
+			}
+
+			// Start batch readers for this generation
+			split, err := r.splitStreams(gen.streams)
+			if err != nil {
+				return err
+			}
+
+			l.Printf("grouped %d streams into %d batches", len(gen.streams), len(split))
+
+			genErrG, genCtx := errgroup.WithContext(runCtx)
+
+			readers := make([]*streamBatchReader, 0, len(split)*len(r.config.TableNames))
+			for _, tableName := range r.config.TableNames {
+				// TODO: This is ugly?
+				splitName := strings.SplitN(tableName, ".", 2)
+				for _, group := range split {
+					readers = append(readers, newStreamBatchReader(
+						r.config,
+						gen.startTime,
+						group,
+						splitName[0],
+						splitName[1],
+						gocql.MinTimeUUID(r.readFrom),
+					))
+				}
+			}
+
+			sleepAmount := r.config.Advanced.PostNonEmptyQueryDelay / time.Duration(len(readers))
+			for i := range readers {
+				reader := readers[i] // TODO: Should this be interruptible?
+				<-time.After(sleepAmount)
+				genErrG.Go(func() error {
+					return reader.run(genCtx)
+				})
+			}
+
+			var nextGen *generation
+			genErrG.Go(func() error {
+				var err error
+				nextGen, err = r.genFetcher.Get(genCtx)
+				if err != nil {
+					return err
+				}
+				for _, reader := range readers {
+					if nextGen == nil {
+						// The reader was stopped
```

Comment:
can the comment explain how we know that?

Reply:
also, you call `stopNow` below, which suggests the possibility that it was not necessarily stopped